### ISSUE TITLE
Using CarveMix on BRATS dataset

### ISSUE THREAD

**Issue by dominikaxx** [ZhangxinruBIT/CarveMix#2]
Hi, I would like to try using CarveMix on BRATS 2021 dataset. Can you please describe the steps I need to follow to make it work with nnU-Net?  You wrote that we need to determine the data partition beforehand, especially for training and validation sets, because validation sets do not participate in CarveMix. How do I do that?

**Comment by ZhangxinruBIT** [comment]
Hi, you can modify at **nnunet.training.net work_training.nnUNetTrainerV2.do_split()**

Specifically, you should make sure the **val_keys** in the  function **nnunet.training.net work_training.nnUNetTrainerV2.do_split()** is the same as **val_set** in **Simple_CarveMix.py**. 
You can directly specify in  **nnunet.training.net work_training.nnUNetTrainerV2.do_split()** such as:

**keys = list(self.dataset.keys())**
**val_keys = ['BraTS2021_XXXX',....]** 
**tr_keys = list(set(keys) - set(val_keys))**

or use **random.seed()** :

**keys = list(self.dataset.keys())**
**raw_keys = [key for key in keys if 'CarveMix' not in key]**
**num_val = int(len(raw_keys)✖️0.2)**
**val_keys = random.sample(raw_keys,num_val)** 
**tr_keys = list(set(keys) - set(val_keys))**


By the way, as for BRATS 2021, the number of the dataset is too large, data augmentation may not work well. 
Considering that the BRATS is a multi-class task, maybe you should try to make λ >0 to make the mask bigger than the whole tumor. Or you can further explore how to be suitable for multi-class tasks.

**Comment by dominikaxx** [comment]
Thank you very much for your advice, it really helped me. 

And can I ask you, why do you think it might not work very well for BRTAS2021? Because the training dataset already contains 1000+ samples and if I create more images by CarveMix augmentation, for example 200 more, the model accuracy wont be better? I am just a beginner and experimenting on my own, so I would be really interested on your point of view, since you are very knowledgeable in this area.

The second question is about making the λ >0 . Where can I set this value, possibly somewhere along these lines? 
`c = np.random.beta(1, 1)#[0,1]             #creat distance

    c = (c-0.5)*2 #[-1.1]
    
    if c>0:
        lam=c*np.min(dis_array)/2              #λl = -1/2|min(dis_array)|
    else:
        lam=c*np.min(dis_array) `

**Comment by ZhangxinruBIT** [comment]
We have not extended to multi-class, this is just what I suggest.
Yes, if you want to modify the range of λ, the code you list is right.

**Comment by PkuClosed** [comment]
> Thank you very much for your advice, it really helped me.
> 
> And can I ask you, why do you think it might not work very well for BRTAS2021? Because the training dataset already contains 1000+ samples and if I create more images by CarveMix augmentation, for example 200 more, the model accuracy wont be better? I am just a beginner and experimenting on my own, so I would be really interested on your point of view, since you are very knowledgeable in this area.
> 
> The second question is about making the λ >0 . Where can I set this value, possibly somewhere along these lines? `c = np.random.beta(1, 1)#[0,1] #creat distance
> 
> ```
> c = (c-0.5)*2 #[-1.1]
> 
> if c>0:
>     lam=c*np.min(dis_array)/2              #λl = -1/2|min(dis_array)|
> else:
>     lam=c*np.min(dis_array) `
> ```

To clarify, I don't think it is accurate to say CarveMix "may not work well". It is simply that as the number of training scans is large, the performance of the model is already good without data augmentation, and the benefit of data augmentation becomes marginal.

**Comment by dominikaxx** [comment]
Hi, I might be wrong, but it seems to me that this code works for multiclass as well. It generated new samples with all 3 tumor subclasses. The condition in code: 
```
 distance = np.where(f, -(dist_func(f, sampling=spacing)),
 dist_func(f, sampling=spacing))
```
is  true for f > 0, so all tumor subclasses are correctly considered and included in carving.

**Comment by PkuClosed** [comment]
> Hi, I might be wrong, but it seems to me that this code works for multiclass as well. It generated new samples with all 3 tumor subclasses. The condition in code:
> 
> ```
>  distance = np.where(f, -(dist_func(f, sampling=spacing)),
>  dist_func(f, sampling=spacing))
> ```
> 
> is true for f > 0, so all tumor subclasses are correctly considered and included in carving.

My understanding is that this is a naive extension to multi-class segmentation. But its performance has not been extensively evaluated. Maybe Xinru can confirm this.
